core/crypto/aes-gcm.c

```diff
@@ -0,0 +1,306 @@
+{
+	size_t n;
+	const uint8_t *s = src;
+	uint8_t *d = dst;
+	size_t l = len;
+
+	if (!ctx->payload_bytes && ctx->buf_pos) {
+		/* AAD part done, finish up the last bits. */
+		memset(ctx->buf_hash + ctx->buf_pos, 0,
+		       TEE_AES_BLOCK_SIZE - ctx->buf_pos);
+		aes_gcm_ghash_update(ctx, ctx->buf_hash, NULL, 0);
+		ctx->buf_pos = 0;
+	}
+
+	ctx->payload_bytes += len;
+
+	while (l) {
+		if (ctx->buf_pos || !ptr_is_block_aligned(s) ||
+		    !ptr_is_block_aligned(d) || l < TEE_AES_BLOCK_SIZE) {
+			n = MIN(TEE_AES_BLOCK_SIZE - ctx->buf_pos, l);
+
+			if (!ctx->buf_pos && m == TEE_MODE_DECRYPT) {
+				aes_gcm_encrypt_block(ctx, ctx->ctr,
+						      ctx->buf_cryp);
+			}
+
+			xor_buf(ctx->buf_cryp + ctx->buf_pos, s, n);
+			memcpy(d, ctx->buf_cryp + ctx->buf_pos, n);
+			if (m == TEE_MODE_ENCRYPT)
+				memcpy(ctx->buf_hash + ctx->buf_pos,
+				       ctx->buf_cryp + ctx->buf_pos, n);
+			else
+				memcpy(ctx->buf_hash + ctx->buf_pos, s, n);
+
+			ctx->buf_pos += n;
+
+			if (ctx->buf_pos != TEE_AES_BLOCK_SIZE)
+				return TEE_SUCCESS;
+
+			aes_gcm_ghash_update(ctx, ctx->buf_hash, NULL, 0);
+			ctx->buf_pos = 0;
+			d += n;
+			s += n;
+			l -= n;
+
+			if (m == TEE_MODE_ENCRYPT)
+				aes_gcm_encrypt_block(ctx, ctx->ctr,
+						      ctx->buf_cryp);
+			aes_gcm_inc_ctr(ctx);
+		}
+
+		if (l > TEE_AES_BLOCK_SIZE &&
+		    ptr_is_block_aligned(s) && ptr_is_block_aligned(d)) {
+			n = l / TEE_AES_BLOCK_SIZE;
+			aes_gcm_update_payload_block_aligned(ctx, m, s, n, d);
+			s += n * TEE_AES_BLOCK_SIZE;
+			d += n * TEE_AES_BLOCK_SIZE;
+			l -= n * TEE_AES_BLOCK_SIZE;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result operation_final(struct aes_gcm_ctx *ctx, TEE_OperationMode m,
+				  const uint8_t *src, size_t len, uint8_t *dst)
+{
+	TEE_Result res;
+
+	res = aes_gcm_update_payload(ctx, m, src, len, dst);
+	if (res)
+		return res;
+
+	if (ctx->buf_pos) {
+		memset(ctx->buf_hash + ctx->buf_pos, 0,
+		       sizeof(ctx->buf_hash) - ctx->buf_pos);
+		aes_gcm_ghash_update(ctx, ctx->buf_hash, NULL, 0);
+	}
+
+	ghash_update_lengths(ctx, ctx->aad_bytes, ctx->payload_bytes);
+	/* buf_tag was filled in with the first counter block aes_gcm_init() */
+	xor_buf(ctx->buf_tag, ctx->hash_state, ctx->tag_len);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result aes_gcm_enc_final(struct aes_gcm_ctx *ctx, const void *src,
+			     size_t len, void *dst, void *tag, size_t *tag_len)
+{
+	TEE_Result res;
+
+	if (*tag_len < ctx->tag_len)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	res = operation_final(ctx, TEE_MODE_ENCRYPT, src, len, dst);
+	if (res)
+		return res;
+
+	memcpy(tag, ctx->buf_tag, ctx->tag_len);
+	*tag_len = ctx->tag_len;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result aes_gcm_dec_final(struct aes_gcm_ctx *ctx, const void *src,
+			     size_t len, void *dst, const void *tag,
+			     size_t tag_len)
+{
+	TEE_Result res;
+
+	if (tag_len != ctx->tag_len)
+		return TEE_ERROR_MAC_INVALID;
+
+	res = operation_final(ctx, TEE_MODE_ENCRYPT, src, len, dst);
```

Comment:
`TEE_MODE_ENCRYPT` => `TEE_MODE_DECRYPT`

Reply:
Thanks, I guess this reveals a blind spot in the test suite.